File: src/main/java/org/tech6/services/pacote/PacoteFotoService.java
package org.tech6.services.pacote;

import org.tech6.dto.pacote.pacoteFoto.PacoteFotoRegistroDTO;
import org.tech6.model.pacote.pacoteFoto.PacoteFoto;
import org.tech6.model.pacote.pacoteFoto.foto.Foto;
import org.tech6.repository.pacote.PacoteFotoRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@ApplicationScoped
public class PacoteFotoService {

    private final PacoteFotoRepository pacoteFotoRepository;

    public PacoteFotoService(PacoteFotoRepository pacoteFotoRepository) {
        this.pacoteFotoRepository = pacoteFotoRepository;
    }

    public List<PacoteFoto> listarPacotesFoto() {
        return pacoteFotoRepository.listAll();
    }

    public PacoteFoto buscarPorId(long id) {
        return pacoteFotoRepository.findById(id);
    }

    @Transactional
    public String criarPacoteFoto(PacoteFotoRegistroDTO dto) {
        PacoteFoto pacoteFoto = new PacoteFoto();
        montarPacoteFoto(pacoteFoto, dto);
        pacoteFotoRepository.persist(pacoteFoto);
        return "Pacote de fotos registrado com sucesso!";
    }

    private void montarPacoteFoto(PacoteFoto pacoteFoto, PacoteFotoRegistroDTO dto) {
        pacoteFoto.nome = dto.nome();
        pacoteFoto.fotoDoPacote = dto.url();

        // Limpa fotos antigas se for edição (o orphanRemoval=true na entidade remove do
        // banco)
        if (pacoteFoto.fotos != null) {
            pacoteFoto.fotos.clear();
        } else {
            pacoteFoto.fotos = new HashSet<>();
        }

        // Adiciona as novas fotos
        if (dto.fotosAdicionais() != null) {
            Set<Foto> fotosEntidade = dto.fotosAdicionais().stream().map(fDto -> {
                Foto f = new Foto();
                f.nome = fDto.nome();
                f.url = fDto.url();
                return f;
            }).collect(Collectors.toSet());

            pacoteFoto.fotos.addAll(fotosEntidade);
        }
    }

    @Transactional
    public String atualizarPacoteFoto(long id, PacoteFotoRegistroDTO dto) {
        PacoteFoto pacoteFoto = pacoteFotoRepository.findById(id);
        if (pacoteFoto == null) {
            return null; // Or throw an exception, depending on desired error handling
        }
        montarPacoteFoto(pacoteFoto, dto);
        // Panache entities are automatically managed within a transaction, no explicit
        // persist() needed for updates
        return "Pacote de fotos atualizado com sucesso!";
    }
}

File: src/main/kotlin/com/fatec/destino/services/pacoteFoto/PacoteFotoService.kt
package com.fatec.destino.services.pacoteFoto

import com.fatec.destino.dto.pacoteFoto.PacoteFotoRegistroDTO
import com.fatec.destino.model.pacote.pacoteFoto.PacoteFoto
import com.fatec.destino.model.pacote.pacoteFoto.foto.Foto
import com.fatec.destino.repository.pacote.pacoteFotoRepository.PacoteFotoRepository
import jakarta.transaction.Transactional
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service

@Service
class PacoteFotoService(
    private val pacoteFotoRepository: PacoteFotoRepository
) {

    fun listarPacotesFoto(): ResponseEntity<List<PacoteFoto>> {
        return ResponseEntity.ok(pacoteFotoRepository.findAll())
    }

    fun buscarPorId(id: Long): ResponseEntity<PacoteFoto> {
        val foto = pacoteFotoRepository.findByIdOrNull(id)
        return if (foto != null) ResponseEntity.ok(foto) else ResponseEntity.notFound().build()
    }

    @Transactional
    fun criarPacoteFoto(dto: PacoteFotoRegistroDTO): ResponseEntity<String> {
        val pacoteFoto = PacoteFoto(
            dto.nome,
            dto.url,
            dto.fotosAdicionais as HashSet<Foto?>
        )
        montarPacoteFoto(pacoteFoto, dto)
        pacoteFotoRepository.save(pacoteFoto)
        return ResponseEntity.ok("Pacote de fotos registrado com sucesso!")
    }

    @Transactional
    fun atualizarPacoteFoto(id: Long, dto: PacoteFotoRegistroDTO): ResponseEntity<String> {
        val pacoteFoto = pacoteFotoRepository.findByIdOrNull(id)
            ?: return ResponseEntity.notFound().build()

        montarPacoteFoto(pacoteFoto, dto)
        pacoteFotoRepository.save(pacoteFoto)
        return ResponseEntity.ok("Pacote de fotos atualizado com sucesso!")
    }

    private fun montarPacoteFoto(pacoteFoto: PacoteFoto, dto: PacoteFotoRegistroDTO) {
        pacoteFoto.nome = dto.nome
        pacoteFoto.fotoDoPacote = dto.url

        pacoteFoto.fotos.clear()

        dto.fotosAdicionais?.forEach { fDto ->
            val novaFoto = Foto().apply {
                nome = fDto.nome
                url = fDto.url
            }
            pacoteFoto.fotos.add(novaFoto)
        }
    }
}